pkg/vendir/cmd/sync.go

```diff
@@ -126,6 +129,31 @@ func (o *SyncOptions) Run() error {
 		if err != nil {
 			return fmt.Errorf("Syncing directory '%s': %s", dirConf.Path, err)
 		}
+		absRoot, err := filepath.Abs(dirConf.Path)
+		if err != nil {
+			return err
+		}
+		if !o.AllowAllSymlinkDestinations {
+			err = filepath.WalkDir(dirConf.Path, func(path string, info fs.DirEntry, err error) error {
+				if info.Type()&os.ModeSymlink == os.ModeSymlink {
+					resolvedPath, err := filepath.EvalSymlinks(path)
+					if err != nil {
+						return fmt.Errorf("Unable to resolve symlink: %w", err)
+					}
+					absPath, err := filepath.Abs(resolvedPath)
+					if err != nil {
+						return err
+					}
+					if !strings.HasPrefix(absPath, absRoot) {
```

Comment:
should we be paranoid and check that it's, either

absPath == absRoot, or
HasPrefix(absPath, absRoot+string(filepath.Separator))

otherwise we could allow pointing to destinations that are "sibling" to allowed dir.

(i did similar logic here: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/pkg/memdir/scoped.go)